crates/binjs_io/src/entropy/model.rs

```diff
@@ -1,183 +1,251 @@
-        Box::new(ExactEncodingModel::new(tree))
+    /// All list lengths, predicted by path.
+    pub list_length_by_path: PathPredict<Option<u32>, T>,
+
+    // Missing:
+    // - offsets (cannot be predicted?)
+    // - property keys predicted by window?
+    // - identifier names predicted by window?
+    // - literal strings by window?
+    // - directives?
+    // - property keys, identifier names, literal strings by file?
+}
+impl<T> Dictionary<T> {
+    /// Return the number of states in this dictionary.
+    pub fn len(&self) -> usize {
+          self.bool_by_path.len()
+        + self.float_by_path.len()
+        + self.unsigned_long_by_path.len()
+        + self.string_enum_by_path.len()
+        + self.property_key_by_path.len()
+        + self.identifier_name_by_path.len()
+        + self.interface_name_by_path.len()
+        + self.string_literal_by_path.len()
+        + self.list_length_by_path.len()
     }
 }
 
+/// Counting instances of each kind of string in a file.
+///
+/// We use this data to determine which strings should go in a shared dictionary
+/// or which should be inlined in the file.
+#[derive(Debug, Default)]
+pub struct StringsInFile<T> {
+    /// Instances of IdentifierName.
+    pub identifier_name_instances: HashMap<Option<IdentifierName>, T>,
+
+    /// Instances of PropertyKey
+    pub property_key_instances: HashMap<Option<PropertyKey>, T>,
+
+    /// Instances of InterfaceName
+    pub interface_name_instances: HashMap<InterfaceName, T>,
+
+    /// Instances of string literals.
+    pub string_literal_instances: HashMap<Option<SharedString>, T>,
 
-/// An encoding model which starts by analyzing the full AST to determine
-/// exact statistics.
-pub struct ExactEncodingModel {
-    probabilities: ExactEncodingModelData<Symbol>,
+    /// Instances of string enums.
+    pub string_enum_instances: HashMap<SharedString, T>,
 }
-impl ExactEncodingModel {
-    fn get_from_path<T>(predictor: &mut PathPredict<T, Symbol>, value: &T, path: &ASTPath) -> Result<Symbol, ()>
-        where T: Eq + std::hash::Hash + Clone + std::fmt::Debug
-    {
-        let segment = predictor.get_mut(path, value)
-            .ok_or(())?;
-        let result = segment.clone();
-        Ok(result)
-    }
+
+/// A structure used to build a dictionary based on a sample of files.
+///
+/// In this version, we use several dictionary lengths, to determine the
+/// best version for a file.
+pub struct DictionaryBuilder<'a> {
+    /// Attempt to build predictions with a depth of 0 (no context) up to max_path_depth ancestors.
+    max_path_depth: usize,
+
+    /// A dictionary.
+    dictionary: &'a mut Dictionary</* instances */ usize>,
+
+    /// Number of instances of each string in the current file.
+    instances_of_strings_in_current_file: StringsInFile<usize>,
+
+    /// Number of files in which each string appears.
+    files_containing_string: &'a mut StringsInFile<usize>,
```

Comment:
can you explain why it's StringsInFile type ?
also, where is the target of reference supposed to be exist ?

Reply:
> can you explain why it's StringsInFile type ?

This should be better now.

>  also, where is the target of reference supposed to be exist ?

Er... anywhere upstack?